File: scripts/validate.py
#!/usr/bin/env python3
"""Run the repository's authoritative validation sequence."""

from __future__ import annotations

import argparse
import subprocess
import sys
from pathlib import Path


def run(package_root: Path, *arguments: str) -> None:
    subprocess.run([sys.executable, *arguments], cwd=package_root, check=True)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--npx", action="store_true", help="include the networked npx-skills installation test")
    args = parser.parse_args()
    package_root = Path(__file__).resolve().parents[1]

    run(package_root, "scripts/build_skill_packages.py", "--check")
    run(package_root, "-m", "pytest")
    for fixture in ("single-proactive", "shared-explicit"):
        run(package_root, "-m", "knowledge_loom.cli", "audit", f"tests/fixtures/{fixture}")
    run(package_root, "scripts/run_behavior_evals.py")
    run(package_root, "scripts/build_claude_desktop_skill.py")
    if args.npx:
        run(package_root, "scripts/test_npx_install.py")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate.ts
import path from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";

import { errorMessage } from "../src/knowledge-loom/errors.ts";

const PACKAGE_ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");

function run(...arguments_: string[]): void {
  const completed = spawnSync(process.execPath, arguments_, { cwd: PACKAGE_ROOT, stdio: "inherit" });
  if (completed.status !== 0) throw new Error(`${arguments_.join(" ")} exited with status ${completed.status}`);
}

export function main(arguments_: string[] = process.argv.slice(2)): number {
  const includeNpx = arguments_.includes("--npx");
  if (arguments_.some((argument) => argument !== "--npx")) throw new Error(`unknown argument: ${arguments_.find((argument) => argument !== "--npx")}`);
  run("node_modules/typescript/bin/tsc", "--noEmit");
  run("--import", "tsx", "scripts/build-skill-packages.ts", "--check");
  run("--import", "tsx", "scripts/run-tests.ts");
  for (const fixture of ["single-proactive", "shared-explicit"]) {
    run("--import", "tsx", "src/knowledge-loom/runner.ts", "audit", `tests/fixtures/${fixture}`);
  }
  run("--import", "tsx", "scripts/run-behavior-evals.ts");
  run("--import", "tsx", "scripts/build-claude-desktop-skill.ts");
  if (includeNpx) run("--import", "tsx", "scripts/test-npx-install.ts");
  return 0;
}

if (path.resolve(process.argv[1] ?? "") === fileURLToPath(import.meta.url)) {
  try { process.exitCode = main(); } catch (error) { console.error(`ERROR ${errorMessage(error)}`); process.exitCode = 1; }
}
